### PR TITLE
put check for DVB duration with pagetimeout

### LINKED ISSUE #1104: [BUG] DVB subtitles timing issue
CCExtractor version (using the --version parameter preferably) : **0.88** (724d756aa6b90daf83e7a652f956ce29c20ac210)

**In raising this issue, I confirm the following (please check boxes, eg [X] - and delete unchecked ones):**

- [X] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [X] I have checked that the bug-fix I am reporting can be replicated, or that the feature I am suggesting isn't already present.
- [X] I have checked that the issue I'm posting isn't already reported.
- [X] I have checked that the issue I'm porting isn't already solved and no duplicates exist in [closed issues](https://github.com/CCExtractor/ccextractor/issues?q=is%3Aissue+is%3Aclosed) and in [opened issues](https://github.com/CCExtractor/ccextractor/issues)
- [X] I have checked the pull requests tab for existing solutions/implementations to my issue/suggestion.
- [X] I have used the latest available version of CCExtractor to verify this issue exists.

**My familiarity with the project is as follows (check one, eg [X] - and delete unchecked ones):**

- [X] I absolutely love CCExtractor, but have not contributed previously.

**Necessary information**
- Is this a regression (did it work before)? [ ] NO | [ ] YES - we havent noticed that before. It only appers to be happening on some channels, not all.
- What platform did you use? [ ] Windows - [X] Linux - [ ] Mac
- What were the used arguments? `ccextractor -in=ts  -ocrlang por -quant 0  -datapid 0x1448  telecine.ts`

**Video links (replace text below with your links)**
https://bit.ly/2kczK9X

**Additional information**
The subtitles are parsed out perfectly, the only issue is that whenever there is silence (nobody speaking), the last subtitle is always shown, until someone starts to speak. You can easily test this in VLC, turn on DVB subs on for example on 0:46, there is a pause in the subtitles (none are showing in VLC). 

In the extracted subtitles, there is no pause, each subtitle is always 1ms after the previous ends - see below:

```
00:00:38.883 --> 00:00:44.987
<font color="#74750b">Por que demoraram tanto?</font>

00:00:44.988 --> 00:00:49.697
<font color="#74750b">Chandler, acaIme-se.</font>
<font color="#74750b">Me deixe preparar minha equipe.</font>
```

Is there a configuration option that we can use to fix this? Or is this an issue in the ccextractor?

Thank you
Jakub

### PR DESCRIPTION
**Pull request by anshul1912**
**In raising this pull request, I confirm the following (please check boxes):**

- [X] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [X] I have checked that another pull request for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.
- [X] **I have mentioned this change in the [changelog](https://github.com/CCExtractor/ccextractor/blob/master/docs/CHANGES.TXT).**

**My familiarity with the project is as follows (check one):**

I was an active contributor to CCExtractor.

---

Possibly fixes #1104

### CODE REVIEW

**Comment by anshul1912** [comment]
Format sourcecode / format having failures in existing files where no change of mine is present, I am not sure if any of mine change is causing this problem

**Comment by canihavesomecoffee** [comment]
@anshul1912 You may ignore that specific check.

I've also linked in the issue you were talking about on Slack.

**Comment by NilsIrl** [comment]
> I am not sure if any of mine change is causing this problem

IFAIK, the timings have never been fully correct

**Comment by canihavesomecoffee** [comment]
> > I am not sure if any of mine change is causing this problem
> 
> IFAIK, the timings have never been fully correct

Anshul was referring to the failing clang-tidy action with his statement ;)

File: src/lib_ccx/dvb_subtitle_decoder.c [hunk 1]
```diff
@@ -1678,12 +1679,17 @@ void dvbsub_handle_display_segment(struct encoder_ctx *enc_ctx,
 				   struct lib_cc_decode *dec_ctx,
 				   struct cc_subtitle *sub)
 {
+  DVBSubContext *ctx = (DVBSubContext *)dec_ctx->private_data;
```
Comment (canihavesomecoffee) [inline src/lib_ccx/dvb_subtitle_decoder.c]:
```suggestion
	DVBSubContext *ctx = (DVBSubContext *)dec_ctx->private_data;
```

File: src/lib_ccx/dvb_subtitle_decoder.c [hunk 2]
```diff
@@ -1678,12 +1679,17 @@ void dvbsub_handle_display_segment(struct encoder_ctx *enc_ctx,
...
 		return;
 	if (enc_ctx->write_previous) //this condition is used for the first subtitle - write_previous will be 0 first so we don't encode a non-existing previous sub
 	{
 		enc_ctx->prev->last_string = NULL;									    // Reset last recognized sub text
 		sub->prev->end_time = (dec_ctx->timing->current_pts - dec_ctx->timing->min_pts) / (MPEG_CLOCK_FREQ / 1000); //we set the end time of the previous sub the current pts
+		if (sub->prev->time_out < sub->prev->end_time - sub->prev->start_time)
+		{
+			 sub->prev->end_time = sub->prev->start_time + sub->prev->time_out;
+		}
 		encode_sub(enc_ctx->prev, sub->prev);									    //we encode it
```
Comment (canihavesomecoffee) [inline src/lib_ccx/dvb_subtitle_decoder.c]:
```suggestion
		encode_sub(enc_ctx->prev, sub->prev); //we encode it

```

File: src/lib_ccx/dvb_subtitle_decoder.c [hunk 3]
```diff
@@ -1678,12 +1679,17 @@ void dvbsub_handle_display_segment(struct encoder_ctx *enc_ctx,
...
+  DVBSubContext *ctx = (DVBSubContext *)dec_ctx->private_data;
 	if (!enc_ctx)
 		return;
 	if (enc_ctx->write_previous) //this condition is used for the first subtitle - write_previous will be 0 first so we don't encode a non-existing previous sub
 	{
 		enc_ctx->prev->last_string = NULL;									    // Reset last recognized sub text
 		sub->prev->end_time = (dec_ctx->timing->current_pts - dec_ctx->timing->min_pts) / (MPEG_CLOCK_FREQ / 1000); //we set the end time of the previous sub the current pts
+		if (sub->prev->time_out < sub->prev->end_time - sub->prev->start_time)
+		{
+			 sub->prev->end_time = sub->prev->start_time + sub->prev->time_out;
```
Comment (canihavesomecoffee) [inline src/lib_ccx/dvb_subtitle_decoder.c]:
```suggestion
			sub->prev->end_time = sub->prev->start_time + sub->prev->time_out;
```

**Comment by canihavesomecoffee** [comment]
@anshul1912 

There's one that seems a bit wonky to me:
ccextractor -autoprogram -out=ttxt -latin1 -datets dcada745de...

https://sampleplatform.ccextractor.org/sample/72

This doesn't seem to be the case yet on master, but I can't spot immediately what changes of you would cause this.

**Comment by anshul1912** [comment]
Verified my change with https://sampleplatform.ccextractor.org/sample/72 there is no change in output with and without it

**Comment by NilsIrl** [comment]
This doesn't fix #1104 for me.

**Comment by NilsIrl** [comment]
> This doesn't fix #1104 for me.

So actually it fixes the time between subtitles but doesn't fix the problem that the start times are wrong. Opening an issue for this different problem.

**Pull request merged**